pennylane/workflow/qnode.py

```diff
@@ -619,22 +611,36 @@ def get_gradient_fn(device, interface, diff_method="best", shots=None):
                 If a string, allowed options are ``"best"``, ``"backprop"``, ``"adjoint"``,
                 ``"device"``, ``"parameter-shift"``, ``"hadamard"``, ``"finite-diff"``, or ``"spsa"``.
                 A gradient transform may also be passed here.
+            tape (Optional[.QuantumTape]): the circuit that will be differentiated. Should include shots information.
 
         Returns:
             tuple[str or .TransformDispatcher, dict, .Device: Tuple containing the ``gradient_fn``,
             ``gradient_kwargs``, and the device to use when calling the execute function.
         """
+
+        config = _make_execution_config(None, diff_method)
+        if isinstance(device, qml.devices.Device):
+            if device.supports_derivatives(config, circuit=tape):
+                new_config = device.preprocess(config)[1]
+                return new_config.gradient_method, {}, device
+            if diff_method in {"backprop", "adjoint", "device"}:  # device-only derivatives
+                raise qml.QuantumFunctionError(
+                    f"Device {device} does not support {diff_method} with requested circuit."

```

Comment:
Won't this error always be raised with `diff_method="device"` since `default.qubit.supports_derivatives` only has branches for "backprop", "adjoint", and "best"?

Reply:
Yes.  default qubit does not actually support `diff_method="device"`.  We technically could start interpreting `"device"` as either backprop or adjoint, but that's not necessarily a feature that's being requested right now.

Reply:
Everything under the new device api.

Reply:
Ah, okay, I was confusing myself. So to add support for `diff_method="device"`, the other devices just need to update `supports_derivatives` so that it returns in the clause above if the method is `"device"`.

Reply:
Yep. `Device.supports_derivatives` just needs to return `True` when `config.gradient_method == "device"`.